File: src/js/xui-one-client.js
/**
 * Minimal XUI.ONE Admin API client.
 * Use only with panels you own or are authorized to manage.
 */
export class XuiOneClient {
  constructor({ baseUrl, apiKey, timeoutMs = 20000 }) {
    if (!baseUrl) throw new Error('baseUrl is required');
    if (!apiKey) throw new Error('apiKey is required');

    this.baseUrl = baseUrl.replace(/\/+$/, '') + '/';
    this.apiKey = apiKey;
    this.timeoutMs = timeoutMs;
  }

  async get(action, params = {}) {
    return this.request('GET', action, params);
  }

  async post(action, params = {}) {
    return this.request('POST', action, params);
  }

  async request(method, action, params = {}) {
    const url = new URL(this.baseUrl);
    url.searchParams.set('api_key', this.apiKey);
    url.searchParams.set('action', action);

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);

    const options = {
      method,
      headers: { Accept: 'application/json' },
      signal: controller.signal
    };

    if (method === 'GET') {
      for (const [key, value] of Object.entries(params)) {
        url.searchParams.set(key, value);
      }
    } else {
      options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
      options.body = new URLSearchParams(params).toString();
    }

    try {
      const response = await fetch(url, options);
      const text = await response.text();

      try {
        return {
          ok: response.ok,
          statusCode: response.status,
          data: JSON.parse(text)
        };
      } catch {
        return {
          ok: response.ok,
          statusCode: response.status,
          raw: text
        };
      }
    } finally {
      clearTimeout(timeout);
    }
  }
}

File: src/python/xui_one_client.py
"""Minimal XUI.ONE Admin API client.

Use only with panels you own or are authorized to manage.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional
from urllib.parse import urlencode
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError
import json
import ssl


@dataclass
class XuiOneClient:
    base_url: str
    api_key: str
    verify_ssl: bool = True
    timeout: int = 20

    def __post_init__(self) -> None:
        self.base_url = self.base_url.rstrip("/") + "/"

    def get(self, action: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self._request("GET", action, params or {})

    def post(self, action: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self._request("POST", action, params or {})

    def _request(self, method: str, action: str, params: Dict[str, Any]) -> Dict[str, Any]:
        query = {"api_key": self.api_key, "action": action}
        url = self.base_url + "?" + urlencode(query)

        body = None
        headers = {"Accept": "application/json"}

        if method.upper() == "GET" and params:
            url += "&" + urlencode(params, doseq=True)
        elif method.upper() == "POST":
            body = urlencode(params, doseq=True).encode("utf-8")
            headers["Content-Type"] = "application/x-www-form-urlencoded"

        context = None
        if not self.verify_ssl:
            context = ssl._create_unverified_context()  # noqa: SLF001

        request = Request(url, data=body, headers=headers, method=method.upper())

        try:
            with urlopen(request, timeout=self.timeout, context=context) as response:
                raw = response.read().decode("utf-8", errors="replace")
                return self._decode(raw, response.status)
        except HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            return self._decode(raw, exc.code)
        except URLError as exc:
            raise RuntimeError(f"Connection error: {exc}") from exc

    @staticmethod
    def _decode(raw: str, status_code: int) -> Dict[str, Any]:
        try:
            data = json.loads(raw)
            return {"ok": 200 <= status_code < 300, "status_code": status_code, "data": data}
        except json.JSONDecodeError:
            return {"ok": 200 <= status_code < 300, "status_code": status_code, "raw": raw}
